File: scripts/seed.py
"""
Embed the catalog + knowledge chunks and load them into Supabase pgvector.

Idempotent: truncates and reloads the two static tables, so re-running after a
re-harvest / re-chunk gives a clean state. Uses a direct Postgres connection
(efficient for bulk vector inserts) and the OpenAI embeddings API.

Prereqs:
  pip install openai "psycopg[binary]"
  export OPENAI_API_KEY=...
  export DATABASE_URL=postgresql://...   # Supabase: Project Settings -> Database

Usage:  python3 scripts/seed.py
"""
import json, os, sys
import psycopg
from openai import OpenAI

DATA = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "data")
MODEL = "text-embedding-3-small"          # 1536-dim; must match the vector() columns
BATCH = 200

client = OpenAI()  # reads OPENAI_API_KEY


def embed_all(texts):
    out = []
    for i in range(0, len(texts), BATCH):
        batch = texts[i:i + BATCH]
        res = client.embeddings.create(model=MODEL, input=batch)
        out.extend(d.embedding for d in res.data)
        print(f"  embedded {min(i + BATCH, len(texts))}/{len(texts)}", flush=True)
    return out


def vec(e):
    return "[" + ",".join(f"{x:.7f}" for x in e) + "]"   # pgvector text literal


def seed_catalog(cur):
    rows = json.load(open(f"{DATA}/catalog.json"))
    texts = [f"{r['title']} — {r.get('label') or ''} ({r.get('agency') or ''})" for r in rows]
    embs = embed_all(texts)
    cur.execute("truncate catalog restart identity")
    with cur.copy(
        "copy catalog (ka, title, label, agency, caid, kasid, entry_url, embedding) from stdin"
    ) as cp:
        for r, e in zip(rows, embs):
            cp.write_row((r["ka"], r["title"], r.get("label"), r.get("agency"),
                          r["caid"], r["kasid"], r["entry_url"], vec(e)))
    print(f"catalog: {len(rows)} rows")


def seed_chunks(cur):
    rows = json.load(open(f"{DATA}/chunks.json"))
    # embed with title + section context so the vector carries topic
    texts = [f"{r['title']}\n{r.get('section') or ''}\n{r['body']}" for r in rows]
    embs = embed_all(texts)
    cur.execute("truncate knowledge_chunks restart identity")
    with cur.copy(
        "copy knowledge_chunks (ka, title, section, body, links, embedding) from stdin"
    ) as cp:
        for r, e in zip(rows, embs):
            cp.write_row((r["ka"], r["title"], r.get("section"), r["body"],
                          json.dumps(r["links"]), vec(e)))
    print(f"knowledge_chunks: {len(rows)} rows")


def main():
    dsn = os.environ.get("DATABASE_URL")
    if not dsn:
        sys.exit("Set DATABASE_URL (Supabase Postgres connection string).")
    with psycopg.connect(dsn) as conn:
        with conn.cursor() as cur:
            print("seeding catalog..."); seed_catalog(cur)
            print("seeding knowledge_chunks..."); seed_chunks(cur)
        conn.commit()
    print("done.")


if __name__ == "__main__":
    main()

File: scripts/seed.mjs
// Seed catalog + knowledge_chunks into Supabase via the REST client (supabase-js).
// No DATABASE_URL needed — uses NEXT_PUBLIC_SUPABASE_URL + SUPABASE_SERVICE_ROLE_KEY.
// Idempotent: clears each table, then bulk-inserts in batches.
//
// Prereq: run the migrations in supabase/migrations/ first (SQL editor).
// Run:    node scripts/seed.mjs
import { readFileSync, existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { createClient } from "@supabase/supabase-js";
import OpenAI from "openai";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const DATA = join(ROOT, "data");

// Minimal .env loader (avoids a Node-version dependency on --env-file).
function loadEnv(path) {
  if (!existsSync(path)) return;
  for (const line of readFileSync(path, "utf8").split("\n")) {
    const m = line.match(/^\s*([\w.]+)\s*=\s*(.*)\s*$/);
    if (!m) continue;
    const v = m[2].replace(/^["']|["']$/g, "");
    if (!(m[1] in process.env)) process.env[m[1]] = v;
  }
}
loadEnv(join(ROOT, ".env"));
loadEnv(join(ROOT, ".env.local"));

const EMBED_MODEL = "text-embedding-3-small";
const EMBED_BATCH = 200; // OpenAI inputs per embeddings request
const INSERT_BATCH = 100; // rows per Supabase insert (keeps payloads small)

const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
const key = process.env.SUPABASE_SERVICE_ROLE_KEY;
if (!url || !key) {
  console.error("Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env");
  process.exit(1);
}
const supabase = createClient(url, key, { auth: { persistSession: false } });
const openai = new OpenAI(); // OPENAI_API_KEY from env

const load = (f) => JSON.parse(readFileSync(join(DATA, f), "utf8"));

async function embedAll(texts) {
  const out = [];
  for (let i = 0; i < texts.length; i += EMBED_BATCH) {
    const res = await openai.embeddings.create({
      model: EMBED_MODEL,
      input: texts.slice(i, i + EMBED_BATCH),
    });
    for (const d of res.data) out.push(d.embedding);
    process.stdout.write(`  embedded ${Math.min(i + EMBED_BATCH, texts.length)}/${texts.length}\r`);
  }
  process.stdout.write("\n");
  return out;
}

async function clearTable(table) {
  const { error } = await supabase.from(table).delete().gte("id", 0);
  if (error) throw new Error(`clear ${table}: ${error.message}`);
}

async function insertAll(table, rows) {
  for (let i = 0; i < rows.length; i += INSERT_BATCH) {
    const { error } = await supabase.from(table).insert(rows.slice(i, i + INSERT_BATCH));
    if (error) throw new Error(`insert ${table} @${i}: ${error.message}`);
    process.stdout.write(`  inserted ${Math.min(i + INSERT_BATCH, rows.length)}/${rows.length}\r`);
  }
  process.stdout.write("\n");
}

async function seedCatalog() {
  const rows = load("catalog.json");
  const texts = rows.map((r) => `${r.title} — ${r.label || ""} (${r.agency || ""})`);
  const embs = await embedAll(texts);
  const records = rows.map((r, i) => ({
    ka: r.ka, title: r.title, label: r.label ?? null, agency: r.agency ?? null,
    caid: r.caid, kasid: r.kasid, entry_url: r.entry_url, embedding: embs[i],
  }));
  await clearTable("catalog");
  await insertAll("catalog", records);
  console.log(`catalog: ${records.length} rows`);
}

async function seedChunks() {
  const rows = load("chunks.json");
  const texts = rows.map((r) => `${r.title}\n${r.section || ""}\n${r.body}`);
  const embs = await embedAll(texts);
  const records = rows.map((r, i) => ({
    ka: r.ka, title: r.title, section: r.section ?? null, body: r.body,
    links: r.links ?? [], embedding: embs[i],
  }));
  await clearTable("knowledge_chunks");
  await insertAll("knowledge_chunks", records);
  console.log(`knowledge_chunks: ${records.length} rows`);
}

console.log("seeding catalog...");
await seedCatalog();
console.log("seeding knowledge_chunks...");
await seedChunks();
console.log("done.");
